barter-data-rs/src/dex/uniswapx/mod.rs

```diff
@@ -1,166 +1,192 @@
-
+use crate::event::{DataKind, MarketEvent};
+use crate::subscription::intent_order::{IntentOrder, IntentOrderUpdate};
+use barter_integration::model::{
+    instrument::{kind::InstrumentKind, Instrument},
+    Exchange,
+};
 use eyre::Result;
 use reqwest;
-use crate::subscription::intent_order::{IntentOrder, IntentOrderUpdate};
 
-use self::uni_order::{UniOrder, Response};
+use self::uni_order::{Response, UniOrder};
+use super::tokens::TokenCache;
 use super::DexError;
-use tokio::time::{Duration, sleep};
 use tokio::sync::mpsc::{self, UnboundedReceiver};
+use tokio::time::{sleep, Duration};
 
 pub mod uni_order;
 
 const UNISWAPX_API: &str = "https://api.uniswap.org/v2/orders";
 
-fn map_uni_orders_to_intent_orders(uni_orders: Vec<UniOrder>, event: IntentOrderUpdate) -> Vec<IntentOrder> {
-  let mut intent_orders = Vec::new();
-
-  for uni_order in uni_orders {
-      // TODO: Fetch decimals for tokens from the token list
-      let start_ask = uni_order.outputs[0].start_amount.parse::<f64>().unwrap() / uni_order.input.start_amount.parse::<f64>().unwrap();
-      let end_ask = uni_order.outputs[0].end_amount.parse::<f64>().unwrap() / uni_order.input.start_amount.parse::<f64>().unwrap();
-      let price = uni_order.outputs[0].end_amount.parse::<f64>().unwrap() / uni_order.input.start_amount.parse::<f64>().unwrap();
-
-      let intent_order = IntentOrder {
-          event,
-          id: uni_order.order_hash.clone(), // You can use any suitable value for id
-          in_token: uni_order.input.token.clone(),
-          in_amount: uni_order.input.start_amount.parse::<f64>().unwrap_or(0.0),
-          out_token: uni_order.outputs[0].token.clone(), // Assuming there is always at least one output
-          out_amount: uni_order.outputs[0].end_amount.parse::<f64>().unwrap_or(0.0),
-          start_ask,
-          end_ask,
-          price,     
-          created_at: uni_order.created_at,
-          order_type: uni_order.order_type.clone(),
-          signature: uni_order.signature.clone(),
-          encoded_order: uni_order.encoded_order.clone(),
-      };
-
-      intent_orders.push(intent_order);
-  }
-
-  intent_orders
+async fn map_uni_orders_to_intent_orders(
+    uni_orders: Vec<UniOrder>,
+    event: IntentOrderUpdate,
+) -> Result<Vec<IntentOrder>, DexError> {
+    let mut intent_orders = Vec::new();
+
+    let tokens = TokenCache::instance().lock().await;
+    for uni_order in uni_orders {
+        // TODO: Why are there multiple output tokens?
+        let token_in = tokens.get_token(&1, &uni_order.input.token).await?;
+        let token_out = tokens.get_token(&1, &uni_order.outputs[0].token).await?;
+        // TODO: Fetch decimals for tokens from the token list
+        let start_ask = uni_order.outputs[0].start_amount.parse::<f64>().unwrap()
```

Comment:
This parsing from string to f64 should be done automatically during deserialization using #[serde(deserialize_with = "barter_integration::de::de_str")] macro in the Uni_Order. (search for example)
 
But we should also consider if these should have Decimal type to ensure precision...